### ISSUE TITLE
Traited Fact vs Fact

### ISSUE THREAD

**Issue by Northburns** [Kryszak/gwatlin#114]
I have some thoughts on Fact vs Traited Fact. I may have some trouble here because I'm so new at the game, but here goes. For reference the [GW2 Wiki's API page on Skills](https://wiki.guildwars2.com/wiki/API:2/skills).

My understanding: There are different kinds of facts, and current implementation of a `sealed interface` + concrete data classes makes a whole lot of sense. Now, some skills may also have a `traited_facts` array of "traited facts" which are _just like_ facts but contain two additional fields (the required trait, and whether its a new fact or overrides existing fact).

> All objects returned within a traited_facts array contain the same properties as a facts object, but they additionally have the following properties: 

So the current implementation of `TraitedFact` seems to need work. I'm wondering what options there are in implementing this. I'd say these would be reasonable goals:

* *(A)* a TraitedFact is a Fact. This way application can calculate the effective chain of facts.
* *(B)* there is a Damage type, whether its a fact or a traited_fact. This way application can handle Damage facts
* *(C)* Concrete classes being data classes is super-nice. Kotlin restriction to take into account: ["Data classes can't be abstract, open, sealed, or inner."](https://kotlinlang.org/docs/data-classes.html)
* *(D)* Sealed types are super-nice. Kotlin Serialization feels the same way: ["The most straightforward way to use serialization with a polymorphic hierarchy is to mark the base class `sealed`."](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md#sealed-classes) (compare to open polymorphism on that page).
* *(E)* Avoiding custom serializers keeps things simpler, but utilizing them is not the end of the world

**Option 1:** Trying to achieve all 5 goals is quite tricky 🤔 Artificial "inbetween interfaces" are messy and would require some explaining 😅 I tried to create some small poc before work today, but I think a custom serializer would need to be created. `JsonContentPolymorphicSerializer` is quite nice to work with, but the Fact type checking would have to be implemented there, instead of using the `@SerialName` - and I also noticed that the type attribute serialization would have to be implemented as well. I've done some work with Kotlin Serialization, but it's quite a long time ago, so I'd need to brush up on the custom serializer side of things 😊

**Option 2:** would be to move the TraitedFact fields (as nullable) to Fact. This would do away with the TraitedFact type, and whether Fact "is traited or not" would be checking if the `requires_trait` is null. This would hide the TraitedFact from types, but would be super simple to implement, at the cost of a bit of added complexity to application.

**Option 3:** How about a custom serializer? There's a bunch of ways this could be done. Maybe a `JsonTransformingSerializer` would be simple enough, if we can make the assumption we'll always handle data as JSON. I created a PR with a proposal #113 .

TL;DR Just read the quoted part and Option 3, sorry for rambling 😅

**Comment by Kryszak** [comment]
Proposed solution is fine to me.
> I may have some trouble here because I'm so new at the game,
I'm afraid is the same case for me because I quit playing GW2 before first expansion came out :smile:

**Comment by Northburns** [comment]
Haha, fair 😄 But it seems a pretty nice API to work with. 

Would it be ok to create a release after this PR? It's not _blocking_ my project, but it'd be make a couple features leap forward 😊

**Comment by Kryszak** [comment]
Yup, I'll prepare and release version `2.0.0` later on.
